src/Orleans.CodeGenerator/CodeGenerator.cs

```diff
@@ -160,6 +160,23 @@ public CompilationUnitSyntax GenerateCode(CancellationToken cancellationToken)
                                     {
                                         constructorParameters = annotatedConstructors[0].Parameters;
                                     }
+                                    else
+                                    {
+                                        // record structs from referenced assemblies do not return IsRecord=true
+                                        // above. See https://github.com/dotnet/roslyn/issues/69326
+                                        // So we implement the same heuristics from ShouldIncludePrimaryConstructorParameters
+                                        // to detect a primary constructor.
+                                        var properties = symbol.GetMembers().OfType<IPropertySymbol>().ToImmutableArray();
+                                        var primaryConstructor = symbol.GetMembers()
+                                            .OfType<IMethodSymbol>()
+                                            .Where(m => m.MethodKind == MethodKind.Constructor && m.Parameters.Length > 0)
+                                            // Check for a ctor where all parameters have a corresponding compiler-generated prop.
+                                            .FirstOrDefault(ctor => ctor.Parameters.All(prm =>
+                                                properties.Any(prop => prop.Name.Equals(prm.Name, StringComparison.Ordinal) && prop.IsCompilerGenerated())));
```

Comment:
How reliable is this? Could some language/compiler change in the future break it and cause a wire compatibility issue for users?

Reply:
I couldn't find anything more reliable than that :(

Reply:
As sharplab shows, there's nothing else we can depend on, unless the runtime adds something: https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEDMACATgUwGMB7XBbAMWOIAp0BGABmwCEJcBKAbiA=. The property itself isn't `[CompilerGenerated]`, and I doubt anyone would manually write an accessor annotated with that.

Reply:
FWIW, I think the proposed heuristics is as specific and detailed as possible to detect the situation given the current compiler output.